registry/remote/repository.go

```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
+	 -----------------------------------------------------------------------------------
+	*/
+	var contentDigest digest.Digest
+
+	// 5a. Set contentDigest if no early error conditions are encountered
+	var calculatedDigest digest.Digest
+	if len(hdrDigest) == 0 {
+		if httpMethod == "HEAD" {
+			// HEAD without server `Docker-Content-Digest` header is an
+			// immediate fail; HEAD
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"HTTP %s request missing required header `%s`",
+				httpMethod, dockerContentDigestHeader,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
 
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+		// GET without server `Docker-Content-Digest` header forces the
+		// expensive calculation
+		if calculatedDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		contentDigest = calculatedDigest
+	} else {
+		contentDigest = hdrDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	// 5b. With all 3 digests appropriately set, perform final 3-way validation
+	if err = verifyContentDigests(calculatedDigest, hdrDigest, refDigest); err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
 	}
 
+	// 6. Finally, if we made it this far, then all is good; return.
 	return ocispec.Descriptor{
 		MediaType: mediaType,
 		Digest:    contentDigest,
 		Size:      size,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body
+// taking care not to destroy it in the process.
+func calculateDigestFromResponse(resp *http.Response) (digest.Digest, error) {
+	// Read from `resp.Body` and (via io.TeeReader), duplicate the read data
+	// into the new buffers.  This read will "deplete" resp.Body which, having
+	// no "rewind" button, will become useless.  At which point, one clone is
+	//used to calculate the digest, and another is used replace the now-defunct
+	// `resp.Body`.
+	var cloneForDigestCalculation io.Reader
+	var cloneForRestoration bytes.Buffer
+
+	var d digest.Digest
+	var err error
+	body := limitReader(resp.Body, 0)
+	cloneForDigestCalculation = io.TeeReader(body, &cloneForRestoration)
+	if d, err = digest.FromReader(cloneForDigestCalculation); err != nil {
+		return d, fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
+	}
+
+	resp.Body = io.NopCloser(&cloneForRestoration)
+
+	return d, nil
```

Comment:
Since we will read all of the response body into the memory anyway, why don't we do something like this?
```go
defer resp.Body.Close() // Yes, we need to close the body to free up resources like connections

body := limitReader(resp.Body, MaxMetadataBytes) // You need to find a way to pass this variable from `Repository` via `manifestStore` or `blobStore`
content, err := io.ReadAll(body)
if err != nil {
    return "", fmt.Errorf("%s %q: failed to read response body: %w", resp.Request.Method, resp.Request.URL, err)
}
resp.Body = io.NopCloser(bytes.NewReader(content))
return digest.FromBytes(content), nil
```

It is much simpler, cleaner, and faster.